design-documents/csp.md

```diff
@@ -0,0 +1,208 @@
+```xml
+<csp_whitelist xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:Security/etc/csp_whitelist.xsd">
+    <policies>
+        <policy id="font-src">
+            <values>
+                <value>fonts.mywebsite.com</value>
+            </values>
+        </policy>
+    </policies>
+</csp_whitelist>
+```
+ 
+#### Extension Developers
+Extension developers will be able to place _csp_whitelist.xml_ inside their modules just as theme developers. For more intricate
+cases they will be able to create a custom _PolicyCollectorInterface_ implementations to edit policies in the runtime.
+ 
+#### Per-page configuration
+It should be possible to configure CSP for a single page to allow access to certain resources/features
+for just that one document. To achieve this developers will be able to implement a special interface by the
+controller for the page they want custom CSP for.
+```php
+namespace Magento\Security\Model\CSP\PolicyCollector;
+
+interface CSPAwareActionInterface extends ActionInterface
+{
+    public function modifyCsp(PolicyInterface[] $appliedPolicies): PolicyInterface[];
+}
+```
+ 
+Thus controllers will be able to exclude/add policies to configured CSPs.
+
+## Reporting
+Browsers can report CSP violations in both modes. Magento will provide a default endpoint to receive these reports.
+Since there is no way to authenticate a genuine report and on a live store they can fill up quickly the number of
+reports stored in the database will be limited to 10000 deleting old ones when the limit is reached.
+ 
+Merchants will see new violations attention message in the admin panel.
+ 
+![alt text](img/csp_violations.png "CSP Violations")
+ 
+Extension developers will be able to redefine default report-uri via config and implement a
+custom _ViolationsRepositoryInterface_ to show them in the admin panel.
+ 
+Violations report should be accessible through web API. Both CSP related admin pages and the web API endpoints
+must require _Magento_Backend::system_ resource access.
+ 
+## Default CSP
+Magento uses some JavaScript libraries but usually has them provided with Magento itself without including
+them from 3rd party resources so for most cases having *-src directives with 'self' value will be enough. In some .phtmls
+we set event listener via attributes and we use eval in template.js so we'd have to also allow
+'unsafe-inline' and 'unsafe-eval'. Magento has integrations with multiple 3rd party services like vimeo, youtube,
+google analytics and various payment systems which we must whitelist additionally.
+That will be the default whitelist Magento provides out of the box packaged alongside related modules via _csp\_whitelist.xml_.
+ 
+## Future steps
+To allow merchants to use whitelisting with _nonce_ we have to get reed of event handlers via HTML attributes and
+style attributes in our templates. There is no way to disable `unsafe-eval` since we use it for UI components and
```

Comment:
> There is no way to disable `unsafe-eval` since we use it for UI components and
some of the front-end libraries we employ need it (like jQuery).

This can be done with a breaking change. If the work is scheduled, it can be done in 2.4

Reply:
As I stated in my comment above referring to this line:

> If we are knowingly using vulnerable libraries and the desire is to avoid the vulnerable feature, we should have an active plan to mitigate the risk.

Reply:
getting reed of eval() in Magento front-end is a separate topic, obviously it has to be done but describing how would take another document.